docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,276 @@
+
+```csharp
+public T Find<T>(IEnumerable<T> sequence, Func<T, bool> match)
+```
+
+You want to continue to express that the sequence doesn't contain `null` values, and the `match` function won't be called with a `null` value. But, if the sought element isn't found, the returned value could be `null`. That's when you add the `MaybeNull` annotation to the method return:
+
+```csharp
+[return: MaybeNull]
+public T Find<T>(IEnumerable<T> sequence, Func<T, bool> match)
+```
+
+The preceding code informs callers that the contract implies a non-nullable type, but the return value *may* actually be null.  Use the `MaybeNull` attribute when your API should be a non-nullable type, typically a generic type parameter, but there may be instances where `null` would be returned.
+
+You can also specify that a return, or an `out` or `ref` argument isn't null even though the type is a nullable type. Consider a method that ensures an `Array` is large enough to hold a number of elements. If the input argument doesn't have capacity, the routine would allocate a new array and copy all the existing elements into it. If the input argument is `null`, the routine would allocate new storage. If there's sufficient capacity, the routine does nothing:
+
+```csharp
+public void EnsureCapacity<T>(ref T[] storage, int size)
+```
+
+You could call this routine as follows:
+
+```csharp
+// messages has the default value (null) when EnsureCapacity is called:
+EnsureCapacity<string>(ref messages, 10);
+// messages is not null.
+EnsureCapacity<string>(messages, 50);
+```
+
+After enabling null reference types, you want to  ensure that the preceding code compiles without warnings. When the method returns, the `storage` argument is guaranteed to be not null. However, it's acceptable to call `EnsureCapacity` with a null reference. You can make `storage` a nullable reference type, and add the `NotNull` post-condition to the parameter declaration:
+
+```csharp
+public void EnsureCapacity<T>([NotNull]ref T[]? storage, int size)
+```
+
+The preceding code expresses the existing contract very clearly: Callers can pass a variable with the `null` value, but the return value is guaranteed to never be null. The `NotNull` attribute is most useful for `ref` and `out` arguments where `null` may be passed as an argument, but that argument is guaranteed to be not null when the method returns.
+
+## Specify conditional post-conditions: `NotNullWhen` and `MaybeNullWhen`
+
+You're likely familiar with the `string` method <xref:System.String.IsNullOrEmpty(string)?DisplayProperty=nameWithType>. This method returns `true` when the argument isn't null, and not the empty string. Callers shouldn't need to null-check the argument if the method returns `false`. To make a method like this nullable aware, you'd set the argument to a nullable type, and add the `NotNullWhen` attribute:
+
+```csharp
+bool IsNullOrEmpty([NotNullWhen(false)]string? value);
```

Comment:
Should this be on the return?

```csharp
[return: NotNullWhen("false")]
bool IsNullOrEmpty(string? value);
```

If not, then something isn't clicking for me since the return value is what is checked on the `if` test below but the attribute is specified on the input argument?

---
**UPDATE**
OK thinking and reading more below is it possible that the next paragraph is actually what threw me off? Why is it that when the return value as `false` would indicate that null wouldn't need to be checked? The compiler wouldn't know if false or true is being returned would it? That is a run-time check.

Reply:
Thanks for noting this. I added more explanation. It is a tough concept.